### PR TITLE
Update GoogleTest to v1.14.0

### PR DESCRIPTION
**Pull request by zeroomega**
This PR updates GoogleTest to v1.14.0
The LLVM specific modification over the GoogleTest can be seen from: https://gist.github.com/zeroomega/fb24b1d1c4252b852200e15fda29384d
List of removed GoogleTest files can be found: 
https://gist.github.com/zeroomega/54d1da74aa5df9a22bf60474adc53cab

2 additional commits are required for the GoogleTest roll. Due to a change in how GoogleTest treats uninstantiated tests, any test added by `TEST_P` needs an corresponding `INSTANTIATE_TEST_SUITE_P`, otherwise an error will be reported. However, in LLVM, sometimes `INSTANTIATE_TEST_SUITE_P` calls were commented out to temporarily disable tests. In this case in v1.14.0, a `GTEST_ALLOW_UNINSTANTIATED_PARAMETERIZED_TEST` is needed to suppress the error. This macro did not exist in v1.10.0 so the change needs to land together with v1.14.0 roll to avoid build breakages.

The roll was tested under Linux x64 and Windows x64 and pass all tests under testing. It would be great if we can tests it under more build variants and hosts in LLVM bots before landing it.

I admit I cannot create a shellscript to do the auto roll. At least it was not possible without some non-trival refactoring to LLVM's patch and move these LLVM's additional functions into `gtest_port.h/cc` and `gmock_port.h/cc`. A better approach would be upstream LLVM's changes to GoogleTest so we can just keep a vanilla GoogleTest in LLVM (using git submodule). 

The GoogleTest code also uses a different code style (Google C++ style instead of LLVM style). Shall I reformat the entire GoogleTest code base?

### CODE REVIEW

**Comment by MaskRay** [comment]
Thank you for the change! `ninja check-llvm check-clang check-clang-tools check-flang check-mlir check-lld check-polly` parses.

> The GoogleTest code also uses a different code style (Google C++ style instead of LLVM style). Shall I reformat the entire GoogleTest code base?

I think we should just use the upstream style.

**Review by MaskRay — APPROVED** [review]

**Review by joker-eph — APPROVED** [review]
Agree with @MaskRay for the style: we should minimize the diff with upstream as much as possible!

File: third-party/unittest/CMakeLists.txt
Comment (petrhosek) [inline third-party/unittest/CMakeLists.txt:81]: This is undoing a change made recently in https://github.com/llvm/llvm-project/commit/91b3ca39667b6341a8c1983a1467fae14b58318b. Is that intentional or was your checkout just stale?

File: third-party/unittest/CMakeLists.txt
Comment (petrhosek) [inline third-party/unittest/CMakeLists.txt:91]: This is undoing a change made recently in https://github.com/llvm/llvm-project/commit/91b3ca39667b6341a8c1983a1467fae14b58318b. Is that intentional or was your checkout just stale?

File: third-party/unittest/googletest/README.LLVM
Comment (petrhosek) [inline third-party/unittest/googletest/README.LLVM:5]: We don't include the revision in the `googlemock` `README.LLVM` file so I'd avoid it here also for consistency (or include it in the other one as well).

File: third-party/unittest/googlemock/README.LLVM
Comment (pogo59) [inline third-party/unittest/googlemock/README.LLVM:19]: Is this still true? If it is, it sounds like the kind of patch that ought to be made in upstream googletest. If not, please remove this line.

File: third-party/unittest/googletest/README.LLVM
Comment (pogo59) [inline third-party/unittest/googletest/README.LLVM:21]: Also include/gtest/gtest-message.h and include/gtest/gtest-printers.h. Probably worth saying something like "(see uses of llvm_gtest)" to help the next person doing a roll.

File: third-party/unittest/googletest/README.LLVM
Comment (pogo59) [reply third-party/unittest/googletest/README.LLVM:5]: Parts of the googletest documentation claim to follow a "live at HEAD" philosophy, so a hash reference (here and in the other README) seems appropriate. Although using a tag, if there is one, would be preferable.

**Review by pogo59 — COMMENTED** [review]
Nice to see the LLVM intrusion into googletest proper is very minimized!

IME, testing on Linux with both gcc and clang as build compilers is very helpful. With Windows/MSVC that covers the three main cases.

File: third-party/unittest/googlemock/README.LLVM
Comment (zeroomega) [reply third-party/unittest/googlemock/README.LLVM:19]: Yes, gmock-matchers.h still have this part patched, see https://gist.github.com/zeroomega/fb24b1d1c4252b852200e15fda29384d#file-gtest_llvm_modification-patch-L251 . When locally test it, without the patch, LLVM unit tests will have build errors.

File: third-party/unittest/googletest/README.LLVM
Comment (zeroomega) [reply third-party/unittest/googletest/README.LLVM:5]: v1.14.0 is a valid tag and `f8d7d77c06936315286eb55f8de22cd23c188571` is the revision that it resolves to. The content matches the 1.14.0 source tar ball. The reason that I include a hash here is because the previous roll v1.10.0 doesn't have a valid tag and I had a bit hard time to find the correct revision for doing a diff to see what LLVM have patched. Since v1.14.0 is a valid tag, I think it is appropriate to remove the hash and just leave v1.14.0 in the doc.

File: third-party/unittest/CMakeLists.txt
Comment (zeroomega) [reply third-party/unittest/CMakeLists.txt:91]: My checkout just stales. Thanks for pointing it out.

File: third-party/unittest/CMakeLists.txt
Comment (zeroomega) [reply third-party/unittest/CMakeLists.txt:81]: My checkout just stales. Thanks for pointing it out.

File: third-party/unittest/googlemock/README.LLVM
Comment (pogo59) [reply third-party/unittest/googlemock/README.LLVM:19]: That seems odd, and worth following up with upstream googletest. In the meantime, I think it would be clearer to future maintainers if we did something to clearly identify the code blocks that were changed, and why. For example,
`// LLVM local change to support std::begin/std::end`
followed by the new code, and then the old code commented-out. That way diffs will provide clues, and merges/commits will have conflicts that someone will need to sort out properly.

File: third-party/unittest/googletest/README.LLVM
Comment (zeroomega) [reply third-party/unittest/googletest/README.LLVM:19]: I added a description for the these two headers.

File: third-party/unittest/googlemock/README.LLVM
Comment (zeroomega) [reply third-party/unittest/googlemock/README.LLVM:19]: I put the local changes into commented blocks in change in amend commit: 2fe51a37bad9fc6aa08f09b4b6f137034e2c8140 in this PR, please take a look.

**Comment by zeroomega** [comment]
I have addressed all review comments in the the latest amend commits. Please take a look.

File: third-party/unittest/googlemock/README.LLVM
Comment (pogo59) [reply third-party/unittest/googlemock/README.LLVM:19]: I looked at that specific commit, seems fine.

**Review by petrhosek — APPROVED** [review]

**Comment by zeroomega** [comment]
I have locally tested this PR using gcc 12.2.0 on Linux x64 and MSVC from VS2019, both passed without issues. @pogo59 Do you have any concerns or suggestions before we merge this?

**Comment by pogo59** [comment]
I'm happy with this, the bots will catch any environmental issues that you haven't found yet. LGTM.

**Comment by zeroomega** [comment]
First bot failure: https://lab.llvm.org/buildbot/#/builders/127/builds/55068

```
[18/126] Generating SANITIZER_TEST_OBJECTS.sanitizer_lzw_test.cpp.x86_64.o
FAILED: projects/compiler-rt/lib/sanitizer_common/tests/SANITIZER_TEST_OBJECTS.sanitizer_lzw_test.cpp.x86_64.o C:/b/slave/sanitizer-windows/build/stage1/projects/compiler-rt/lib/sanitizer_common/tests/SANITIZER_TEST_OBJECTS.sanitizer_lzw_test.cpp.x86_64.o 
cmd.exe /C "cd /D C:\b\slave\sanitizer-windows\build\stage1\projects\compiler-rt\lib\sanitizer_common\tests && C:\b\slave\sanitizer-windows\build\stage1\.\bin\clang.exe -DGTEST_NO_LLVM_SUPPORT=1 -DGTEST_HAS_RTTI=0 -IC:/b/slave/sanitizer-windows/llvm-project/llvm/../third-party/unittest/googletest/include -IC:/b/slave/sanitizer-windows/llvm-project/llvm/../third-party/unittest/googletest -Wno-deprecated-declarations -DGTEST_NO_LLVM_SUPPORT=1 -DGTEST_HAS_RTTI=0 -IC:/b/slave/sanitizer-windows/llvm-project/llvm/../third-party/unittest/googlemock/include -IC:/b/slave/sanitizer-windows/llvm-project/llvm/../third-party/unittest/googlemock -IC:/b/slave/sanitizer-windows/llvm-project/compiler-rt/include -IC:/b/slave/sanitizer-windows/llvm-project/compiler-rt/lib -IC:/b/slave/sanitizer-windows/llvm-project/compiler-rt/lib/sanitizer_common -DSANITIZER_COMMON_NO_REDEFINE_BUILTINS -fno-rtti -O2 -Werror=sign-compare -Wno-gnu-zero-variadic-macro-arguments -gline-tables-only -gcodeview -c -o SANITIZER_TEST_OBJECTS.sanitizer_lzw_test.cpp.x86_64.o C:/b/slave/sanitizer-windows/llvm-project/compiler-rt/lib/sanitizer_common/tests/sanitizer_lzw_test.cpp"
C:/b/slave/sanitizer-windows/llvm-project/compiler-rt/lib/sanitizer_common/tests/sanitizer_lzw_test.cpp:22:10: error: no member named 'generate' in namespace 'std'
   22 |     std::generate(data.begin(), data.end(), gen);
      |     ~~~~~^
```
Probably caused by missing include header. Looking into it.

**Comment by zeroomega** [comment]
> I'm seeing some unused functions in files that haven't been changed in years, which I'm guessing come from this update - did the new googletest change how custom output works?
> 
> ```
> llvm/unittests/Support/JSONTest.cpp:490:27: error: unused function 'operator<<' [-Werror,-Wunused-function]
> inline llvm::raw_ostream &operator<<(llvm::raw_ostream &OS,
>                           ^
> 1 error generated.
> ```
> 
> I guess we can just delete this function, but I'm not sure if it means we've degraded the error messages for these tests or anything like that.

Yes new gtest changes how customized function work. These functions either need to be deleted or revised.

I plan to revert this roll as I am seeing some linker error in bots:

```
FAILED: compiler-rt/lib/fuzzer/tests/FuzzerUtils-aarch64-Test /b/sanitizer-aarch64-linux-fuzzer/build/llvm_build0/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/tests/FuzzerUtils-aarch64-Test
cd /b/sanitizer-aarch64-linux-fuzzer/build/llvm_build0/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/tests && /b/sanitizer-aarch64-linux-fuzzer/build/llvm_build0/./bin/clang++ FuzzedDataProviderTestObjects.FuzzedDataProviderUnittest.cpp.aarch64.o FuzzedDataProviderTestObjects.gtest-all.cc.aarch64.o -o /b/sanitizer-aarch64-linux-fuzzer/build/llvm_build0/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/tests/./FuzzerUtils-aarch64-Test -fuse-ld=lld -Wl,--color-diagnostics -Wthread-safety -Wthread-safety-reference -Wthread-safety-beta --driver-mode=g++ -lpthread -nostdlib++ -fno-exceptions /b/sanitizer-aarch64-linux-fuzzer/build/llvm_build0/runtimes/runtimes-bins/compiler-rt/lib/fuzzer/libcxx_fuzzer_aarch64/lib/libc++.a -march=armv8-a
ld.lld: error: undefined symbol: testing::internal2::PrintBytesInObjectTo(unsigned char const*, unsigned long, std::__Fuzzer::basic_ostream<char, std::__Fuzzer::char_traits<char>>*)
>>> referenced by gtest-printers.h:158 (/b/sanitizer-aarch64-linux-fuzzer/build/llvm-project/runtimes/../third-party/unittest/googletest/include/gtest/gtest-printers.h:158)
>>>               FuzzedDataProviderTestObjects.FuzzedDataProviderUnittest.cpp.aarch64.o:(std::__Fuzzer::basic_string<char, std::__Fuzzer::char_traits<char>, std::__Fuzzer::allocator<char>> testing::internal::FormatForComparisonFailureMessage<FuzzedDataProvider_ConsumeEnum_Test::TestBody()::Enum, FuzzedDataProvider_ConsumeEnum_Test::TestBody()::Enum>(FuzzedDataProvider_ConsumeEnum_Test::TestBody()::Enum const&, FuzzedDataProvider_ConsumeEnum_Test::TestBody()::Enum const&))
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

It will need some time to debug.

**Comment by zeroomega** [comment]
I looked into the "unused functions" issue shown up in bots like: https://lab.llvm.org/buildbot/#/builders/57/builds/29853 .

The root cause is that GoogleTest changed the way it converts an arbitrary object into string. In the past, any object that cannot be printed to the string will eventually matched to function at https://github.com/llvm/llvm-project/blob/58d50b0cadafe118faf2e7d7bf738d2daa38bb73/third-party/unittest/googletest/include/gtest/gtest-printers.h#L264 . Then the call to the llvm_gtest::printable will use `OS << V.V;` to print the value to the stream, if the object overrides its `<<` operator, like https://github.com/llvm/llvm-project/blob/58d50b0cadafe118faf2e7d7bf738d2daa38bb73/clang-tools-extra/include-cleaner/unittests/AnalysisTest.cpp#L457 , it will be used to convert the object to string. If not, a build error will happen. In GoogleTest v1.14.0, the object to string logic is re-written and for an arbitrary  object, it will eventually matched to `RawBytesPrinter` defined at https://github.com/llvm/llvm-project/blob/54c1a9b20d89e85cd60d002c77b34c00f36520f4/third-party/unittest/googletest/include/gtest/gtest-printers.h#L290 . Therefore, the `operator<<` override won't be used and a unused function warnig(error) will be thrown. 

So deleting these `llvm::raw_ostream &operator<<` functions won't cause any test failure but it will degrade the error message output since now the objects were printed as raw bytes. @bogner FYI.

**Comment by zeroomega** [comment]
Relanded as a866ce789eb99da4d7a486eeb60a53be6c75f4fd. I will monitor the bots and manually trigger clean build if linker error appears again.

**Comment by mikaelholmen** [comment]
I see compilation problems with this patch.
By mistake I originally posted about the problems in
https://github.com/llvm/llvm-project/issues/66268#issuecomment-1718791073

So the problem: On a RHEL7 server we see stuff like:
`
In file included from /repo/uabelho/dev-main/compiler-rt/lib/memprof/tests/driver.cpp:9:
In file included from /repo/uabelho/dev-main/runtimes/../third-party/unittest/googletest/include/gtest/gtest.h:65:
In file included from /repo/uabelho/dev-main/runtimes/../third-party/unittest/googletest/include/gtest/gtest-death-test.h:43:
In file included from /repo/uabelho/dev-main/runtimes/../third-party/unittest/googletest/include/gtest/internal/gtest-death-test-internal.h:47:
/repo/uabelho/dev-main/runtimes/../third-party/unittest/googletest/include/gtest/gtest-matchers.h:728:50: error: too few template arguments for class template 'equal_to'
  728 |       : ComparisonBase<EqMatcher<Rhs>, Rhs, std::equal_to<>>(rhs) {}
      |                                                  ^
/usr/lib/gcc/x86_64-redhat-linux/4.8.5/../../../../include/c++/4.8.5/bits/stl_function.h:204:12: note: template is declared here
  203 |   template<typename _Tp>
      |   ~~~~~~~~~~~~~~~~~~~~~~
  204 |     struct equal_to : public binary_function<_Tp, _Tp, bool>
      |            ^
`
so it looks like it's when it's using the newly built clang to compile 
 compiler-rt/lib/memprof/tests/driver.cpp
it's using header files from the host and those headers are too old for the updated GoogleTest.

If we don't build builtins (LLVM_BUILTIN_TARGETS) I don't see any problems.

**Comment by mikaelholmen** [comment]
Ok, seems like the only problem is the MemProf unit test
 compiler-rt/lib/memprof/tests/

So if I manually disable that I don't see any other failures with this patch.

**Comment by zeroomega** [comment]
> Ok, seems like the only problem is the MemProf unit test compiler-rt/lib/memprof/tests/
> 
> So if I manually disable that I don't see any other failures with this patch.

I think any components that uses freshly built clang (stage1) could potentially run into this problem again, the correct way to solve it is to provide a compatible C++ Library to the freshly built clang when it built compiler-rt and related tests. For example, use -D with `RUNTIMES_${target}_CMAKE_CXX_FLAGS` and `RUNTIMES_${target}_CMAKE_CXX_FLAGS` to pass `--gcc-toolchain` (it might be deprecated and there is an updated flag for it) to the runtimes build. This will probably allow stage1 clang to use gcc's stdlib like your bootstrap clang.

**Comment by bjope** [comment]
> > Ok, seems like the only problem is the MemProf unit test compiler-rt/lib/memprof/tests/
> > So if I manually disable that I don't see any other failures with this patch.
> 
> I think any components that uses freshly built clang (stage1) could potentially run into this problem again, the correct way to solve it is to provide a compatible C++ Library to the freshly built clang when it built compiler-rt and related tests. For example, use -D with `RUNTIMES_${target}_CMAKE_CXX_FLAGS` and `RUNTIMES_${target}_CMAKE_CXX_FLAGS` to pass `--gcc-toolchain` (it might be deprecated and there is an updated flag for it) to the runtimes build. This will probably allow stage1 clang to use gcc's stdlib like your bootstrap clang.

Thanks! Yes, I think we hadn't considered that some things we set up for the regular build wasn't forwarded to the compiler-rt test.
When setting
  `-DRUNTIMES_${target}_COMPILER_RT_TEST_COMPILER_CFLAGS=--gcc-toolchain=<path-to-toolchain>`
the problem @mikaelholmen mentioned disappeared. I haven't tried adding it to CMAKE_CXX_FLAGS, but maybe that is a better option.

**Comment by zmodem** [comment]
We're seeing unit test crashes on Windows when building with rpmalloc (via `LLVM_INTEGRATED_CRT_ALLOC`). It's crashing here: https://github.com/google/googletest/blob/v1.14.0/googletest/include/gtest/internal/gtest-port.h#L2092

+ @aganea have you also hit this?

(Our bug: https://crbug.com/1487548)

**Comment by aganea** [comment]
@zmodem Can you add `ENABLE_ASSERTS` on [this line](https://github.com/llvm/llvm-project/blob/720e3bacbd9fdba05645a4d621d43ad712c44df3/llvm/lib/Support/CMakeLists.txt#L107) and run the test again, see if that gives something interesting?

**Comment by zmodem** [comment]
Just tried it, but I didn't hit any asserts.

**Comment by petrhosek** [comment]
We have seen occasional segfaults when using rpmalloc in other tools such as GN and Ninja which we haven't been able to get to the bottom of so we eventually switched to another allocator. It's possible that it's the same issue here, in which case perhaps we should reconsider our support for rpmalloc.

**Comment by aganea** [comment]
@zmodem I haven’t been able to repro with the recipe posted on https://crbug.com/1487548. I am using latest rpmalloc (main branch), latest VS 2022, latest WinSDK. I am running on Win11, on a AMD Ryzen9 CPU. Are you running on WinServer2022? Are able to give more precision on the conditions for reproducing?

**Comment by zmodem** [comment]
I believe we're on the 10.0.19041.0 (2020-04) SDK and VS 16.6.1, running on Windows 10. The rpmalloc version is bc1923f which is probably old by now.

I'll try with the latest rpmalloc version, and will also see if I can dig into the stl headers a bit, but I won't be able to get to this today.

**Comment by aganea** [comment]
I managed to repro. It is actually this, still open, issue: https://github.com/microsoft/STL/issues/1066

**Pull request merged**
